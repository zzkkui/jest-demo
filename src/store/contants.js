export const CHANGE_INPUT_VALUE = 'changeInputValue'

export const CHANGE_ITEMS = 'changeItems'