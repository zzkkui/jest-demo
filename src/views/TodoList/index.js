import React, { Component } from 'react'
import { connect } from 'react-redux'
import Header from './components/Header'
import UndoList from './components/UndoList'
import DoList from './components/DoList'
import './index.css'

export class TodoList extends Component {
  render() {
    const { undoItems } = this.props;
    return (
      <div>
        <Header />
        <UndoList />
        <DoList list={undoItems} />
      </div>
    )
  }
}

export default connect(
  ({ todo }) => ({ undoItems: todo.undoItems })
)(TodoList)
