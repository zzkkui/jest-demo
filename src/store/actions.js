import { CHANGE_INPUT_VALUE, CHANGE_ITEMS } from './contants'

export const changeInputValue = value => ({
  type: CHANGE_INPUT_VALUE,
  value
})

export const changeItems = value => ({
  type: CHANGE_ITEMS,
  value
})