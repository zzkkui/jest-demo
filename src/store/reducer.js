import { CHANGE_INPUT_VALUE, CHANGE_ITEMS } from './contants'

const initState = {
  inputValue: '',
  undoItems: []
}

export default (state = initState, action) => {
  const { type, value } = action
  // const newState = {...state}
  switch (type) {
    case CHANGE_INPUT_VALUE:
      return { ...state, inputValue: value }
    case CHANGE_ITEMS:
      return { ...state, undoItems: value }
    default:
      return state
  }

}