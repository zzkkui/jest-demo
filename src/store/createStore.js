import { createStore, combineReducers } from 'redux'
import reducer from './reducer'
import * as actions from './actions'

const reducers = combineReducers({ todo: reducer })

const store = createStore(reducers, window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__())

export { actions, reducer }

export default store