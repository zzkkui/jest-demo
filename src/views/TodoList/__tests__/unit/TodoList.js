import React from 'react';
import { Provider } from 'react-redux'
import { shallow, mount } from 'enzyme';
import TodoList from '../../index';
import store from '../../../../store/createStore'

// jest.mock("react-redux", () => ({
//   connect(/* mapStateToProps, mapDispatchToProps */) {
//     return function (WrappedComponent) {
//       function FakeConnect() {
//         return <WrappedComponent />;
//       }
//       const name = WrappedComponent.displayName || WrappedComponent.name || "Component";
//       FakeConnect.displayName = `Connect(${name})`;
//       return FakeConnect;
//     };
//   }
// }));

const list = [{
  value: 'jest',
  isFocus: true,
  isChecked: false
}, {
  value: 'TDD',
  isFocus: false,
  isChecked: false
}, {
  value: 'react',
  isFocus: false,
  isChecked: false
}]

describe('TodoList 组件', () => {
  it('样式渲染正常', () => {
    const warpper = shallow(<TodoList store={store} />).find('TodoList').dive()
    expect(warpper).toMatchSnapshot()
  })

  // it('初始化列表为空', () => {
  //   const warpper = shallow(<Provider store={store}><TodoList /></Provider>)
  //   // console.log(warpper.prop('value').store.getState().todo.undoItems)
  //   expect((warpper.prop('value').store.getState().todo.undoItems)).toEqual([])
  // })

  it('Header 组件存在 changeInputValue、changeItems、inputValue、undoItems 属性', () => {
    // const warpper = mount(shallow(<Provider store={store}><TodoList /></Provider>).get(0))
    // //包括 redux 注入的 props
    // const ele = warpper.find('Header')
    // // 不包括 redux 注入的 props
    // // const ele = warpper.find('Connect(Header)')
    // console.log(ele.props())

    const warpper = mount(<Provider store={store}><TodoList /></Provider>)
    const ele = warpper.find('Header')
    // const ele = warpper.find('Connect(Header)')
    // console.log(ele.props())
    expect(ele.prop('changeInputValue')).toBeTruthy()
    expect(ele.prop('changeItems')).toBeTruthy()
    expect(ele.prop('inputValue')).toBe('')
    expect(ele.prop('undoItems')).toEqual([])

    //## 浅渲染， 只能 find 组件，不能 find 高阶组件
    // 需要 mock react-redux  不包括 redux 注入的 props
    // const warpper = shallow(<TodoList />).first().shallow()
    // const ele = warpper.find('Connect(Header)')
    // console.log(ele.props())

    // 不包括 redux 注入的 props
    // const warpper = shallow(<TodoList store={store} />).find('TodoList').dive()
    // const ele = warpper.find('Connect(Header)')
    // console.log(ele.props())
  })

  it('DoList 组件存在 list 属性', () => {
    const warpper = shallow(<TodoList store={store} />).find('TodoList').dive()
    const DoList = warpper.find('DoList')
    expect(DoList.prop('list')).toEqual([])
  })
});

