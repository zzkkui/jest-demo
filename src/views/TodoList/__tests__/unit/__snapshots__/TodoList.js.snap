// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList 组件 样式渲染正常 1`] = `
<div>
  <Connect(Header) />
  <Connect(UndoList) />
  <DoList
    list={Array []}
  />
</div>
`;
